File: app/src/main/java/com/sergioloc/hologram/Adapter/RecyclerAdapter.java
package com.sergioloc.hologram.Adapter;

import android.content.Context;
import android.content.Intent;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import com.chauthai.swipereveallayout.SwipeRevealLayout;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.sergioloc.hologram.Fragments.ListFragment;
import com.sergioloc.hologram.Models.VideoModel;
import com.sergioloc.hologram.R;
import com.sergioloc.hologram.Activities.PlayerActivity;
import com.varunest.sparkbutton.SparkButton;
import com.varunest.sparkbutton.SparkEventListener;

import java.util.ArrayList;

/**
 * Created by Sergio López on 14/09/2017.
 */

public class RecyclerAdapter extends RecyclerView.Adapter<RecyclerAdapter.MyViewHolder> {

    public static final int SPAN_COUNT_ONE = 1;
    public static final int SPAN_COUNT_THREE = 3;

    private static final int VIEW_TYPE_BOX = 1;
    private static final int VIEW_TYPE_LIST = 2;
    private static final int VIEW_TYPE_LIST_FAV = 3;
    private static final int VIEW_TYPE_BOX_FAV = 4;


    public static boolean FAV_LIST = false;

    public static ArrayList<VideoModel> fav_list;
    public static ArrayList<Integer> fav_id;
    private SwipeRevealLayout lastSwipeLayout;

    //Firebase
    private FirebaseDatabase database;
    private DatabaseReference userFav, userHide;
    private FirebaseUser user;


    private ArrayList<VideoModel> array;
    private GridLayoutManager mLayoutManager;
    private Context context;
    private Boolean guest;

    public RecyclerAdapter(ArrayList<VideoModel> array, GridLayoutManager layoutManager, Boolean guest){
        this.array=array;
        this.mLayoutManager=layoutManager;
        this.guest = guest;
    }

    @Override
    public int getItemViewType(int position) {
        int spanCount = mLayoutManager.getSpanCount();
        if (spanCount == SPAN_COUNT_ONE) {
            if(FAV_LIST){
                return VIEW_TYPE_LIST_FAV;
            }else{
                return VIEW_TYPE_LIST;
            }
        } else {
            if (FAV_LIST){
                return VIEW_TYPE_BOX_FAV;
            }else{
                return VIEW_TYPE_BOX;
            }
        }
    }

    @Override
    public MyViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {

        View view;

        if (viewType == VIEW_TYPE_LIST) {
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_video, parent, false);
        } else if(viewType == VIEW_TYPE_LIST_FAV){
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_video_fav, parent, false);
        } else if(viewType == VIEW_TYPE_BOX){
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_video_box, parent, false);
        } else {
            view = LayoutInflater.from(parent.getContext()).inflate(R.layout.card_video_box_fav, parent, false);
        }
        context = view.getContext();
        return new MyViewHolder(view, viewType);

    }

    @Override
    public void onBindViewHolder(final MyViewHolder holder, final int position) {
        int id = context.getResources().getIdentifier(array.get(position).getImage(), "drawable", context.getPackageName());
        int name = context.getResources().getIdentifier(array.get(position).getName(), "string", context.getPackageName());
        holder.image.setImageResource(id);
        holder.text.setText(context.getResources().getString(name));

        if(guest){

            holder.shieldFav.setVisibility(View.VISIBLE);
            holder.shieldHide.setVisibility(View.VISIBLE);

            holder.shieldFav.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    Toast.makeText(context, "Debes estar registrado para añadir a favoritos",Toast.LENGTH_SHORT).show();
                }
            });
            holder.shieldHide.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    Toast.makeText(context, "Debes estar registrado para ocultar",Toast.LENGTH_SHORT).show();
                }
            });

            if (holder.type == 1 || holder.type == 2){
                holder.swipe_layout.close(true);
                colorTags(holder,position);
            }

        }else{
            if(holder.type == 1 || holder.type == 2)
                colorTags(holder,position);

            initFav();

            if (holder.type == 1 || holder.type == 3){ // Lista general
                loadInfo(holder,position);
            }else { // Lista favoritos
                loadInfoFav(holder,position);
            }

            if(holder.type == 1 || holder.type == 2)
                holder.swipe_layout.close(true);

            holder.swipe_layout.setSwipeListener(new SwipeRevealLayout.SwipeListener() {
                @Override
                public void onClosed(SwipeRevealLayout view) {

                }

                @Override
                public void onOpened(SwipeRevealLayout view) {
                    if (lastSwipeLayout == null){
                        lastSwipeLayout = holder.swipe_layout;
                    }else {
                        lastSwipeLayout.close(true);
                        lastSwipeLayout = holder.swipe_layout;
                    }
                }

                @Override
                public void onSlide(SwipeRevealLayout view, float slideOffset) {
                }
            });


        }

        holder.button.setOnClickListener(new View.OnClickListener(){
            @Override
            public void onClick(View v) {
                Intent i = new Intent(v.getContext(),PlayerActivity.class);
                i.putExtra("id",ListFragment.actualList.get(position).getCode());
                context.startActivity(i);
            }
        });



    }

    /**Firebase**/

    public void initFav(){
        database = FirebaseDatabase.getInstance();
        user = FirebaseAuth.getInstance().getCurrentUser();
        userFav = database.getReference("users").child(user.getUid()).child("fav");
        userHide = database.getReference("users").child(user.getUid()).child("hide");
    }
    public void addFav(int position){
        String name;
        int id = ListFragment.actualList.get(position).getId();
        if(id<10)
            name="video00"+id;
        else if(id<100)
            name="video0"+id;
        else
            name="video"+id;
        userFav.child(name).setValue(ListFragment.actualList.get(position));
    }
    public void removeFav(int position){
        String name;
        int id = ListFragment.actualList.get(position).getId();
        if(id<10)
            name="video00"+id;
        else if(id<100)
            name="video0"+id;
        else
            name="video"+id;
        userFav.child(name).removeValue();
    }
    public void addHide(int position){
        String name;
        int id = ListFragment.actualList.get(position).getId();
        if(id<10)
            name="video00"+id;
        else if(id<100)
            name="video0"+id;
        else
            name="video"+id;
        int nameSnackbar = context.getResources().getIdentifier(ListFragment.actualList.get(position).getName(), "string", context.getPackageName());
        showSnackbar(context.getResources().getString(nameSnackbar),id,name);
        userHide.child(name).setValue(ListFragment.actualList.get(position));
    }
    public void removeHide(int position){
        String name;
        int id = ListFragment.actualList.get(position).getId();
        if(id<10)
            name="video00"+id;
        else if(id<100)
            name="video0"+id;
        else
            name="video"+id;
        userHide.child(name).removeValue();
    }


    /**Views**/

    private void loadInfo(final MyViewHolder holder, final int position){
        holder.bFav.setChecked(false);
        userFav.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                    VideoModel video = snapshot.getValue(VideoModel.class);

                    int name = context.getResources().getIdentifier(video.getName(), "string", context.getPackageName());
                    if(holder.text.getText().toString().equals(context.getResources().getString(name))){
                        holder.bFav.setChecked(true);
                    }
                }
            }
            @Override
            public void onCancelled(DatabaseError databaseError) {
                Toast.makeText(context, databaseError.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });

        // Favorito
        holder.bFav.setEventListener(new SparkEventListener() {
            @Override
            public void onEvent(ImageView button, boolean buttonState) {
                if(buttonState){ //active
                    Toast.makeText(context,"Añadido a favoritos",Toast.LENGTH_SHORT).show();
                    addFav(position);
                }else{ //inactive
                    Toast.makeText(context,"Eliminado de favoritos",Toast.LENGTH_SHORT).show();
                    removeFav(position);
                }
            }

            @Override
            public void onEventAnimationEnd(ImageView button, boolean buttonState) {

            }

            @Override
            public void onEventAnimationStart(ImageView button, boolean buttonState) {

            }
        });

        // Ocultar
        holder.bHide.setEventListener(new SparkEventListener() {
            @Override
            public void onEvent(ImageView button, boolean buttonState) {
                if(buttonState){ //active
                    Toast.makeText(context,"Añadido a ocultos",Toast.LENGTH_SHORT).show();
                    addHide(position);
                }else{ //inactive
                    Toast.makeText(context,"Eliminado de ocultos",Toast.LENGTH_SHORT).show();
                    removeHide(position);
                }
            }

            @Override
            public void onEventAnimationEnd(ImageView button, boolean buttonState) {

            }

            @Override
            public void onEventAnimationStart(ImageView button, boolean buttonState) {

            }
        });

    }
    private void loadInfoFav(final MyViewHolder holder, final int position){
        holder.bFav.setChecked(true);
        holder.bFav.setEventListener(new SparkEventListener() {
            @Override
            public void onEvent(ImageView button, boolean buttonState) {
                if(buttonState){ //active
                    Toast.makeText(context,"Añadido a favoritos",Toast.LENGTH_SHORT).show();
                    addFav(position);
                }else{ //inactive
                    Toast.makeText(context,"Eliminado de favoritos",Toast.LENGTH_SHORT).show();
                    removeFav(position);
                }
            }
            @Override
            public void onEventAnimationEnd(ImageView button, boolean buttonState) {

            }

            @Override
            public void onEventAnimationStart(ImageView button, boolean buttonState) {

            }
        });
    }



    /**Functions**/

    private void colorTags(final MyViewHolder holder, final int position){
        if(array.get(position).getTag().equals("Animals")){
            holder.tag.setBackground(context.getResources().getDrawable(R.drawable.circle_orange));
        }else if(array.get(position).getTag().equals("Films")){
            holder.tag.setBackground(context.getResources().getDrawable(R.drawable.circle_blue));
        }else if(array.get(position).getTag().equals("Space")){
            holder.tag.setBackground(context.getResources().getDrawable(R.drawable.circle_pink));
        }else if(array.get(position).getTag().equals("Natur")){
            holder.tag.setBackground(context.getResources().getDrawable(R.drawable.circle_green));
        }else if(array.get(position).getTag().equals("Music")){
            holder.tag.setBackground(context.getResources().getDrawable(R.drawable.circle_cyan));
        }else if(array.get(position).getTag().equals("Figures")){
            holder.tag.setBackground(context.getResources().getDrawable(R.drawable.circle_yellow));
        }else if(array.get(position).getTag().equals("Others")){
            holder.tag.setBackground(context.getResources().getDrawable(R.drawable.circle_purple));
        }
    }
    private void showSnackbar(String text, final int id, final String name){
        Snackbar snackbar = Snackbar
                .make(ListFragment.recyclerView, text+" hide", 3000)
                .setAction("UNDO", new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        userHide.addValueEventListener(new ValueEventListener() {
                            @Override
                            public void onDataChange(DataSnapshot dataSnapshot) {
                                for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                                    VideoModel video = snapshot.getValue(VideoModel.class);
                                    if(video.getId()==id)
                                        userHide.child(name).removeValue();
                                }
                            }

                            @Override
                            public void onCancelled(DatabaseError databaseError) {

                            }
                        });
                    }
                });

        snackbar.show();
    }
    private void animOut(final MyViewHolder holder){
        float translationX;
        translationX = holder.swipe_layout.getTranslationX() == 0 ? holder.swipe_layout.getWidth() : 0;
        holder.swipe_layout.animate().translationX(-translationX).start();
    }


    public void setFilter(ArrayList<VideoModel> newList){
        array= new ArrayList<>();
        array.addAll(newList);
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return array.size();
    }


    //----------------------------------------------------------------------------------------------
    // Internal class

    public static class MyViewHolder extends RecyclerView.ViewHolder{

        ImageView image, tag, shieldFav, shieldHide;
        TextView text;
        Button button;
        int type;
        SwipeRevealLayout swipe_layout;
        SparkButton bFav,bHide;

        public MyViewHolder(final View itemView, int viewType) {
            super(itemView);
            fav_list = new ArrayList();
            fav_id = new ArrayList<>();
            if (viewType == VIEW_TYPE_LIST) {
                image = (ImageView) itemView.findViewById(R.id.image_big);
                text = (TextView) itemView.findViewById(R.id.title_big);
                button = (Button) itemView.findViewById(R.id.button_big);
                bFav = itemView.findViewById(R.id.button_fav_big);
                bHide = itemView.findViewById(R.id.button_hide_big);
                shieldFav = (ImageView) itemView.findViewById(R.id.button_fav_big_copy);
                shieldHide = (ImageView) itemView.findViewById(R.id.button_hide_big_copy);
                tag = (ImageView) itemView.findViewById(R.id.iv_tag);
                swipe_layout = (SwipeRevealLayout) itemView.findViewById(R.id.swipe_layout_big);
                type=1;
            }else if (viewType == VIEW_TYPE_LIST_FAV){
                image = (ImageView) itemView.findViewById(R.id.image_list_fav);
                text = (TextView) itemView.findViewById(R.id.title_list_fav);
                button = (Button) itemView.findViewById(R.id.button_list_fav);
                bFav = itemView.findViewById(R.id.button_fav_list);
                tag = (ImageView) itemView.findViewById(R.id.iv_tag_fav);
                swipe_layout = (SwipeRevealLayout) itemView.findViewById(R.id.swipe_layout_list_fav);
                type=2;
            }else if (viewType == VIEW_TYPE_BOX){
                image = (ImageView) itemView.findViewById(R.id.image_small);
                text = (TextView) itemView.findViewById(R.id.title_small);
                button = (Button) itemView.findViewById(R.id.button_small);
                bFav = itemView.findViewById(R.id.button_fav_small);
                bHide = itemView.findViewById(R.id.button_hide_small);
                shieldFav = (ImageView) itemView.findViewById(R.id.ivShieldFavBox);
                shieldHide = (ImageView) itemView.findViewById(R.id.ivShieldHideBox);
                type=3;
            }else {
                image = (ImageView) itemView.findViewById(R.id.image_small);
                text = (TextView) itemView.findViewById(R.id.title_small);
                button = (Button) itemView.findViewById(R.id.button_small);
                bFav = itemView.findViewById(R.id.button_fav_small);
                type=4;
            }
        }
    }

}

File: app/src/main/java/com/sergioloc/hologram/Adapter/RecyclerAdapter.kt
package com.sergioloc.hologram.Adapter

import android.content.Context
import android.content.Intent
import android.os.Handler
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.chauthai.swipereveallayout.SwipeRevealLayout
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.database.*
import com.sergioloc.hologram.Interfaces.CatalogInterface
import com.sergioloc.hologram.Models.VideoModel
import com.sergioloc.hologram.Views.PlayerActivity
import com.varunest.sparkbutton.SparkButton
import com.varunest.sparkbutton.SparkEventListener
import java.util.ArrayList
import android.view.animation.AnimationUtils
import com.bumptech.glide.Glide
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.sergioloc.hologram.Models.FirebaseImage
import com.sergioloc.hologram.R


class RecyclerAdapter(var array: ArrayList<VideoModel>, var layoutManager: GridLayoutManager, var guest: Boolean,
                      var interactor: CatalogInterface.Interactor, var presenter: CatalogInterface.Presenter):
        RecyclerView.Adapter<RecyclerAdapter.MyViewHolder>() {

    var fav_list: ArrayList<VideoModel>? = null
    var fav_id: ArrayList<Int>? = null

    var FAV_LIST = false
    var lastSwipeLayout: SwipeRevealLayout? = null

    // Firebase
    private var database: FirebaseDatabase? = null
    private var mStorage: StorageReference? = null
    private var userFav: DatabaseReference? = null
    private var user: FirebaseUser? = null

    // View types
    private val VIEW_TYPE_LIST = 1
    private val VIEW_TYPE_LIST_FAV = 3

    private var context: Context? = null
    private var lastPosition = -1


    override fun getItemViewType(position: Int): Int {
        return if (FAV_LIST) {
                VIEW_TYPE_LIST_FAV
            } else {
                VIEW_TYPE_LIST
            }

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        if (!guest)
            initFirebase()
        val view: View = when (viewType) {
            VIEW_TYPE_LIST -> LayoutInflater.from(parent.context).inflate(R.layout.card_video, parent, false)
            else -> LayoutInflater.from(parent.context).inflate(R.layout.card_video_fav, parent, false)
        }
        context = view.context
        fav_list = ArrayList()
        fav_id = ArrayList()
        return MyViewHolder(view, viewType)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        //val id = context?.resources?.getIdentifier(array[position].image, "drawable", context?.packageName)

        holder.text?.text = array[position].name

        var path = "gs://hologram-2.appspot.com/images/catalog/" + array[position].name + ".png"
        mStorage = FirebaseStorage.getInstance().getReferenceFromUrl(path)
        mStorage?.downloadUrl?.addOnSuccessListener { uri ->
            Glide.with(context!!)
                    .load(uri)
                    .into(holder.image!!)
        }

        val animation = AnimationUtils.loadAnimation(context,
                if (position > lastPosition)
                    R.anim.up_from_bottom
                else
                    R.anim.down_from_top)
        holder.itemView.startAnimation(animation)
        lastPosition = position


        if (guest){
            holder.shieldFav?.visibility = View.VISIBLE
            holder.shieldHide?.visibility = View.VISIBLE
            holder.shieldFav?.setOnClickListener { Toast.makeText(context, R.string.registered_fav, Toast.LENGTH_SHORT).show() }
        }
        else{ //User
            if (holder.type == VIEW_TYPE_LIST) {
                loadList(holder, position)
            } else {
                loadFavList(holder, position)
            }
        }

        setColorTags(holder, position)
        holder.swipeLayout?.close(true)
        holder.swipeLayout?.setSwipeListener(object : SwipeRevealLayout.SwipeListener {
            override fun onClosed(view: SwipeRevealLayout) {}

            override fun onOpened(view: SwipeRevealLayout) {
                if (lastSwipeLayout == null) {
                    lastSwipeLayout = holder.swipeLayout
                } else {
                    lastSwipeLayout?.close(true)
                    lastSwipeLayout = holder.swipeLayout
                }
            }

            override fun onSlide(view: SwipeRevealLayout, slideOffset: Float) {}
        })

        holder.button?.setOnClickListener { v ->
            val i = Intent(v.context, PlayerActivity::class.java)
            i.putExtra("id", interactor.getVideo(position)?.code)
            context?.startActivity(i)
        }
    }

    override fun getItemCount(): Int { return array.size }


    /** Firebase **/

    private fun initFirebase(){
        database = FirebaseDatabase.getInstance()
        user = FirebaseAuth.getInstance().currentUser
        userFav = database?.getReference("users")?.child(user?.uid)?.child("fav")
    }

    private fun addVideoToFav(position: Int){
        userFav?.child(getVideoName(position))?.setValue(interactor.getVideo(position))
    }

    private fun removeVideoFromFav(position: Int){
        presenter?.startLoadingFavList()
        userFav?.child(getVideoName(position))?.removeValue()
        Handler().postDelayed(
                {
                    interactor?.loadFavList()
                    Handler().postDelayed({ presenter?.finishLoadingFavList() }, 20)
                },
                20
        )
    }

    private fun getVideoName(position: Int): String{
        val name: String
        val id = interactor.getVideo(position)!!.id
        name = when {
            id < 10 -> "video00$id"
            id < 100 -> "video0$id"
            else -> "video$id"
        }
        return name
    }

    /** Views **/

    private fun loadList(holder: MyViewHolder, position: Int){
        holder.bFav?.isChecked = false
        userFav?.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                for (snapshot in dataSnapshot.children) {
                    val video = snapshot.getValue(VideoModel::class.java)
                    if (holder.text?.text.toString() == video!!.name) {
                        holder.bFav?.isChecked = true
                    }
                }
            }
            override fun onCancelled(databaseError: DatabaseError) {
                Toast.makeText(context, databaseError.message, Toast.LENGTH_SHORT).show()
            }
        })

        // Fav button
        holder.bFav?.setEventListener(object : SparkEventListener {
            override fun onEvent(button: ImageView, buttonState: Boolean) {
                if (buttonState) { //active
                    Toast.makeText(context, R.string.add_favourites, Toast.LENGTH_SHORT).show()
                    addVideoToFav(position)
                } else { //inactive
                    Toast.makeText(context, R.string.delete_favourites, Toast.LENGTH_SHORT).show()
                    removeVideoFromFav(position)
                }
            }
            override fun onEventAnimationEnd(button: ImageView, buttonState: Boolean) {}
            override fun onEventAnimationStart(button: ImageView, buttonState: Boolean) {}
        })
    }

    private fun loadFavList(holder: MyViewHolder, position: Int){
        holder.bFav?.isChecked = true
        holder.bFav?.setEventListener(object : SparkEventListener {
            override fun onEvent(button: ImageView, buttonState: Boolean) {
                if (buttonState) { //active
                    Toast.makeText(context, R.string.add_favourites, Toast.LENGTH_SHORT).show()
                    addVideoToFav(position)
                } else { //inactive
                    Toast.makeText(context, R.string.delete_favourites, Toast.LENGTH_SHORT).show()
                    removeVideoFromFav(position)
                }
            }
            override fun onEventAnimationEnd(button: ImageView, buttonState: Boolean) {}
            override fun onEventAnimationStart(button: ImageView, buttonState: Boolean) {}
        })
    }


    /** Functions **/

    private fun setColorTags(holder: MyViewHolder, position: Int) {
        when {
            array[position].tag == "Animals" -> holder.tag?.background = context?.resources?.getDrawable(R.drawable.circle_orange)
            array[position].tag == "Films" -> holder.tag?.background = context?.resources?.getDrawable(R.drawable.circle_blue)
            array[position].tag == "Space" -> holder.tag?.background = context?.resources?.getDrawable(R.drawable.circle_pink)
            array[position].tag == "Nature" -> holder.tag?.background = context?.resources?.getDrawable(R.drawable.circle_green)
            array[position].tag == "Music" -> holder.tag?.background = context?.resources?.getDrawable(R.drawable.circle_cyan)
            array[position].tag == "Figures" -> holder.tag?.background = context?.resources?.getDrawable(R.drawable.circle_yellow)
            array[position].tag == "Others" -> holder.tag?.background = context?.resources?.getDrawable(R.drawable.circle_purple)
        }
    }

    fun setFilter(newList: ArrayList<VideoModel>) {
        array = ArrayList()
        array.addAll(newList)
        notifyDataSetChanged()
    }

    fun closeLastSwipeLayout(){
        lastSwipeLayout?.close(true)
    }

    /** Holder Class **/

    class MyViewHolder(itemView: View, viewType: Int): RecyclerView.ViewHolder(itemView) {

        // View types
        val VIEW_TYPE_LIST = 1
        val VIEW_TYPE_LIST_FAV = 3

        var image: ImageView? = null
        var tag: ImageView? = null
        var shieldFav: ImageView? = null
        var shieldHide: ImageView? = null
        var text: TextView? = null
        var button: Button? = null
        var type: Int = 0
        var swipeLayout: SwipeRevealLayout? = null
        var bFav: SparkButton? = null

        init {
            when (viewType) {
                VIEW_TYPE_LIST -> {
                    showListView()
                }
                VIEW_TYPE_LIST_FAV -> {
                    showListFavView()
                }
            }
        }

        private fun showListView(){
            image = itemView.findViewById(R.id.image_big) as ImageView
            text = itemView.findViewById(R.id.title_big) as TextView
            button = itemView.findViewById(R.id.button_big) as Button
            bFav = itemView.findViewById(R.id.button_fav_big)
            shieldFav = itemView.findViewById(R.id.button_fav_big_copy) as ImageView
            tag = itemView.findViewById(R.id.iv_tag) as ImageView
            swipeLayout = itemView.findViewById(R.id.swipe_layout_big) as SwipeRevealLayout
            type = 1
        }
        private fun showListFavView(){
            image = itemView.findViewById(R.id.image_list_fav) as ImageView
            text = itemView.findViewById(R.id.title_list_fav) as TextView
            button = itemView.findViewById(R.id.button_list_fav) as Button
            bFav = itemView.findViewById(R.id.button_fav_list)
            tag = itemView.findViewById(R.id.iv_tag_fav) as ImageView
            swipeLayout = itemView.findViewById(R.id.swipe_layout_list_fav) as SwipeRevealLayout
            type = 3
        }
    }
}